### PR TITLE
POC JWT token implementation

### PR DESCRIPTION
**Pull request by cr0mbly**
### Summary

Issue #209

This Pull request is our fist attempt at implementing a JWT approach for authentication for both Corpora and the MI API we want provide access too.

This is built off the excellent [simplejwt](https://django-rest-framework-simplejwt.readthedocs.io/) package that integrates directly into django/DRF.

The lifecycle for access will be this:

1. User submits Username/Password to `<domain>/mi-token-auth/` and on sucessful auth will be returned with a access and refresh JWT token, E.G.

```CURL
curl --location --request POST 'http://172.28.128.4:8081/mi-token-auth/' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'username=admin' \
--data-urlencode 'password=admin'
```

2. This token can then be stored in the MI API cache and submitted to the MI API for timestamp validation
3. Once the expiery period has lapsed of three hours a new token will need to be accessed via the `<domain>/mi-token-auth-refresh/`

NOTE: URL names any/timeouts and of the role claim names are all subject to change for clarity so feel free to pitch suggestions.

### Deserialization in python

Here's a quick script that works in pythin using the  [PyJWT](https://pyjwt.readthedocs.io/en/stable/) package

```Python
token = '<our-jwt-token>'
public_key = b'<our-public-rsa-key>'
decoded = jwt.decode(token, public_key, algorithms=["RS256"])
decoded
>>> {
  "token_type": "access",
  "exp": 1615844679,
  "jti": "f08aade87f644fbfb31b882e2751f5dd",
  "user_id": 1,
  "iss": "koreromaori.com",
  "sub": "admin",
  "roles": [
    "add_recordingmetadata",
    "add_recordingqualitycontrol",
    "change_text"
  ]
}
```

This script will return the `Claims` when the time set (3 hours) hasn't lapsed but will raise an exception requiring a new token generated when it has, which we can catch on the client APIs end.

### Role claims

The Role claims are just based on the prebuilt [django permissions](https://docs.djangoproject.com/en/3.1/topics/auth/default/#permissions-and-authorization) e.g.

image

Which will allow us to play around with the roles via the django admin but this is just for expediency and we can adjust this and tail it to what we need later on.

### Security

The Public and Private RSA keys for this are stored in an encrypted vault.yml Ansible file and loaded into the environment during build. DM me for the public key so you can test the above.

### Other changes
Changlog:
 - Cleaned up existing urls.py/settings.py
 - Fixed Ipython environment on local
 - Re-enabled sentry for tracking errors.

### CODE REVIEW

File: corpora-deploy/roles/web/tasks/setup_django_app.yml [hunk 1]
```diff
@@ -130,18 +123,6 @@
...
-- name: Create Django Superuser
-  shell: . {{ virtualenv_path }}/bin/activate
-         && . {{ virtualenv_path }}/bin/postactivate
-         && cd {{ application_path }}
-         && echo "from django.db import DEFAULT_DB_ALIAS as database; from django.contrib import auth; model = auth.get_user_model(); model.objects.db_manager(database).create_superuser('{{django_superuser_name}}','{{django_superuser_email}}', '{{django_superuser_password}}')" | python3 manage.py shell
-  when: (create_django_superuser or rebuild_database) and server_type != 'media'
-  tags:
-    - django
-    - rebuild-database
-  ignore_errors: yes
```
Comment (cr0mbly) [inline corpora-deploy/roles/web/tasks/setup_django_app.yml:143]: This is broken and doesn't run during setup and will soon be replaced by #213

File: corpora-deploy/roles/web/templates/virtualenv_postactivate.j2 [hunk 2]
```diff
@@ -75,3 +75,7 @@ export JWT_SECRET="{{jwt_secret}}"
 export NOTEBOOK_DIR="{{ notebook_path }}"
 
 export FILE_ENCRYPTION_KEY="{{file_encryption_key}}"
+
+
+export JWT_AUTH_SSL_PRIVATE="{{JWT_AUTH_SSL_PRIVATE}}"
+export JWT_AUTH_SSL_PUBLIC="{{JWT_AUTH_SSL_PUBLIC}}"
```
Comment (cr0mbly) [inline corpora-deploy/roles/web/templates/virtualenv_postactivate.j2:81]: Injected from the `vault.yml` file

File: corpora/corpora/base_settings.py [hunk 3]
```diff
@@ -594,19 +577,39 @@
...
+    'AUTH_HEADER_TYPES': ('Bearer',),
+    'AUTH_HEADER_NAME': 'HTTP_AUTHORIZATION',
+    'USER_ID_FIELD': 'id',
+    'USER_ID_CLAIM': 'user_id',
+    'AUTH_TOKEN_CLASSES': ('rest_framework_simplejwt.tokens.AccessToken',),
+    'TOKEN_TYPE_CLAIM': 'token_type',
+    'JTI_CLAIM': 'jti',
+    'SLIDING_TOKEN_REFRESH_EXP_CLAIM': 'refresh_exp',
+    'SLIDING_TOKEN_LIFETIME': timedelta(minutes=5),
+    'SLIDING_TOKEN_REFRESH_LIFETIME': timedelta(days=1),
```
Comment (cr0mbly) [inline corpora/corpora/base_settings.py:614]: @timClicks here's the base configuration I havn't adjusted much apart from what you mentioned re times. You can see the full breakdown  [here](https://django-rest-framework-simplejwt.readthedocs.io/en/latest/settings.html) If you feel anything else needs adjusting

File: corpora/corpora/settings.py [hunk 4]
```diff
@@ -1,9 +1,3 @@
-
-try:
-    from .local_settings import *
```
Comment (cr0mbly) [inline corpora/corpora/settings.py:3]: there's no `local_settings` in app so :scissors:

**Comment by timClicks** [comment]
Hey Aiden, I get back into the office tomorrow so will get to this review
tomorrow

File: corpora/corpora/base_settings.py (hunk 3 above)
Comment (timClicks) [reply corpora/corpora/base_settings.py:614]: Sliding tokens are currently mis-configured. We need to adjust `AUTH_TOKEN_CLASSES`.

> If you want to use sliding tokens, change the AUTH_TOKEN_CLASSES setting to ('rest_framework_simplejwt.tokens.SlidingToken',). (Alternatively, the AUTH_TOKEN_CLASSES setting may include dot paths to both the AccessToken and SlidingToken token classes in the rest_framework_simplejwt.tokens module if you want to allow both token types to be used for authentication.)
>
> &mdash; https://django-rest-framework-simplejwt.readthedocs.io/en/latest/token_types.html

I'm also not sure about what the significance of having `ACCESS_TOKEN_LIFETIME` and `SLIDING_TOKEN_LIFETIME` different is...

File: corpora/corpora/serializers/jwt_authentication.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+ISS_DOMAIN = 'koreromaori.com'
+
+class MiApiObtainPairSerializer(TokenObtainPairSerializer):
+    @classmethod
+    def get_token(cls, user):
+        token = super().get_token(user)
+        token['iss'] = ISS_DOMAIN
+        token['sub'] = user.username
+        token['roles'] = list(
```
Comment (timClicks) [inline corpora/corpora/serializers/jwt_authentication.py:12]: Do we want to have separate claims per API? Probably a question for @kmahelona and @utunga

**Review by timClicks — APPROVED** [review]
Code looks good.

I would put the two URL views as `token/` and `token/refresh`, but that's a pretty minor consideration.

Deferring to @kmahelona about auth policy.

File: corpora/corpora/base_settings.py (hunk 3 above)
Comment (cr0mbly) [reply corpora/corpora/base_settings.py:614]: Cool thanks nice catch! I've reset the config to the plugin defaults for all the other unconfigured settings.

File: corpora/corpora/base_settings.py [hunk 6]
```diff
@@ -594,19 +577,39 @@
...
-    #     dsn="https://ba4e6615d22748dc98442c7c29696e87@o232677.ingest.sentry.io/5458955",
-    #     integrations=[DjangoIntegration()],
-    #     traces_sample_rate=1.0,
-    #     environment=ENV_TYPE,
-    #     # If you wish to associate users to errors (assuming you are using
-    #     # django.contrib.auth) you may enable sending PII data.
-    #     send_default_pii=True
-    # )
+    sentry_sdk.init(
+        dsn="https://ba4e6615d22748dc98442c7c29696e87@o232677.ingest.sentry.io/5458955",
```
[the remark this answers is not in the document]
Comment (cr0mbly) [reply corpora/corpora/base_settings.py:583]: Cool thing with the sentry plugin is we only need a single url for all the environments we can map staging/production to different "buckets" using the `environment=ENV_TYPE` kwarg.

File: corpora/corpora/serializers/jwt_authentication.py (hunk 5 above)
Comment (cr0mbly) [reply corpora/corpora/serializers/jwt_authentication.py:12]: yeah I figure this is a point of discussion, keen to get this live as is and we can discuss further our approaches for roles going forward

Comment (timClicks) [reply corpora/corpora/serializers/jwt_authentication.py:12]: Makes perfect sense.

**Pull request merged**